File: next/kmp/core/src/commonMain/kotlin/org/dweb_browser/core/ipc/helper/IpcStreamData.kt
package org.dweb_browser.core.ipc.helper

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import org.dweb_browser.helper.OrderBy
import org.dweb_browser.helper.base64String
import org.dweb_browser.helper.utf8String

@Serializable()
@SerialName(IPC_MESSAGE_TYPE_STREAM_DATA)
data class IpcStreamDataRawString(
  val stream_id: String,
  val encoding: IPC_DATA_ENCODING,
  val data: String,
  override val order: Int,
) : IpcRawMessage, OrderBy {
  fun toIpcStreamData() = IpcStreamData(stream_id, encoding, data, order)
  override fun toString(): String {
    return "IpcStreamDataRawString(stream_id=$stream_id, encoding=$encoding, order=$order, data=${
      when (val len = data.length) {
        in 0..100 -> data
        else -> data.slice(0..19) + "..." + data.slice(len - 20..<len)
      }
    })"
  }
}


@Serializable()
@SerialName(IPC_MESSAGE_TYPE_STREAM_DATA)
data class IpcStreamDataRawBinary(
  val stream_id: String,
  val encoding: IPC_DATA_ENCODING,
  val data: ByteArray,
  override val order: Int,
) : IpcRawMessage, OrderBy {
  fun toIpcStreamData() = IpcStreamData(stream_id, encoding, data, order)
}


data class IpcStreamData(
  override val stream_id: String,
  val encoding: IPC_DATA_ENCODING,
  val data: Any, /*String or ByteArray*/
  override val order: Int,
) : IpcMessage, IpcStream, RawAble<IpcRawMessage>, OrderBy {

  override fun toString(): String {
    return "IpcStreamData(stream_id=$stream_id, encoding=$encoding, order=$order, data=${
      when (data) {
        is ByteArray -> "ByteArray[size=${data.size}]"
        is String -> when (val len = data.length) {
          in 0..100 -> data
          else -> data.slice(0..20) + "..." + data.slice(len - 20..<len)
        }

        else -> data.toString()
      }
    })"
  }

  companion object {
    fun fromBinary(streamId: String, data: ByteArray, order: Int = streamId.hashCode()) =
      IpcStreamData(streamId, IPC_DATA_ENCODING.BINARY, data, order)

    fun fromBase64(streamId: String, data: ByteArray, order: Int = streamId.hashCode()) =
      IpcStreamData(streamId, IPC_DATA_ENCODING.BASE64, data.base64String, order)

    fun fromUtf8(streamId: String, data: ByteArray, order: Int = streamId.hashCode()) =
      fromUtf8(streamId, data.utf8String, order)

    fun fromUtf8(streamId: String, data: String, order: Int = streamId.hashCode()) =
      IpcStreamData(streamId, IPC_DATA_ENCODING.UTF8, data, order)
  }

  val binary by lazy {
    dataToBinary(data, encoding)
  }

  val text by lazy {
    dataToText(data, encoding)
  }

  override val stringAble by lazy {
    when (encoding) {
      IPC_DATA_ENCODING.BINARY -> IpcStreamDataRawString(
        stream_id, IPC_DATA_ENCODING.BASE64, (data as ByteArray).base64String, order
      )

      IPC_DATA_ENCODING.BASE64, IPC_DATA_ENCODING.UTF8 -> IpcStreamDataRawString(
        stream_id, encoding, data as String, order
      )
    }
  }
  override val binaryAble by lazy {
    when (encoding) {
      IPC_DATA_ENCODING.BINARY -> IpcStreamDataRawBinary(
        stream_id, encoding, data as ByteArray, order
      )

      IPC_DATA_ENCODING.BASE64, IPC_DATA_ENCODING.UTF8 -> IpcStreamDataRawBinary(
        stream_id, IPC_DATA_ENCODING.BINARY, binary, order
      )
    }
  }
}

File: toolkit/dweb-core/src/ipc/ipc-message/stream/IpcStreamData.ts
import type { OrderBy } from "@dweb-browser/helper/OrderBy.ts";
import { $once } from "@dweb-browser/helper/decorator/$once.ts";
import { simpleDecoder } from "@dweb-browser/helper/encoding.ts";
import { stringHashCode } from "@dweb-browser/helper/hashCode.ts";
import { CUSTOM_INSPECT } from "@dweb-browser/helper/logger.ts";
import { $dataToBinary, $dataToText, IPC_DATA_ENCODING } from "../internal/IpcData.ts";
import { IPC_MESSAGE_TYPE, ipcMessageBase } from "../internal/IpcMessage.ts";

export type $IpcStreamData = ReturnType<typeof ipcStreamData>;
const ipcStreamData = (stream_id: string, data: string | Uint8Array, encoding: IPC_DATA_ENCODING, order: number) =>
  ({ ...ipcMessageBase(IPC_MESSAGE_TYPE.STREAM_DATA), stream_id, data, encoding, order } as const satisfies OrderBy);
export const IpcStreamData = Object.assign(ipcStreamData, {
  fromBase64(
    stream_id: string,
    data: Uint8Array,
    order: number = stringHashCode(stream_id)
  ): { [key: string | symbol]: unknown } {
    return Object.assign(ipcStreamData(stream_id, simpleDecoder(data, "base64"), IPC_DATA_ENCODING.BASE64, order), {
      // 打印的时候只会输出前20位和后20位，中间以 "..." 省略
      [CUSTOM_INSPECT]() {
        return JSON.stringify({
          ...this,
          data: data.length <= 100 ? data : `${data.slice(0, 20)}...${data.slice(-20)}`,
        });
      },
    });
  },
  fromBinary(stream_id: string, data: Uint8Array, order: number = stringHashCode(stream_id)) {
    return Object.assign(ipcStreamData(stream_id, data, IPC_DATA_ENCODING.BINARY, order), {
      toJSON: $once(() => IpcStreamData.fromBase64(stream_id, data, order)),
    });
  },
  fromUtf8(stream_id: string, data: Uint8Array, order: number = stringHashCode(stream_id)) {
    return ipcStreamData(stream_id, simpleDecoder(data, "utf8"), IPC_DATA_ENCODING.UTF8, order);
  },
  binary(streamData: $IpcStreamData) {
    return $dataToBinary(streamData.data, streamData.encoding);
  },
  text(streamData: $IpcStreamData) {
    return $dataToText(streamData.data, streamData.encoding);
  },
});
